compiler/rustc_lint/src/context.rs

```diff
@@ -419,29 +419,49 @@ impl LintStore {
         }
     }
 
-    fn no_lint_suggestion(&self, lint_name: &str) -> CheckLintNameResult<'_> {
+    fn no_lint_suggestion(&self, lint_name: &str, tool_name: &str) -> CheckLintNameResult<'_> {
         let name_lower = lint_name.to_lowercase();
 
         if lint_name.chars().any(char::is_uppercase) && self.find_lints(&name_lower).is_ok() {
             // First check if the lint name is (partly) in upper case instead of lower case...
-            return CheckLintNameResult::NoLint(Some(Symbol::intern(&name_lower)));
+            return CheckLintNameResult::NoLint(Some((Symbol::intern(&name_lower), false)));
         }
+
         // ...if not, search for lints with a similar name
-        // Note: find_best_match_for_name depends on the sort order of its input vector.
-        // To ensure deterministic output, sort elements of the lint_groups hash map.
         // Also, never suggest deprecated lint groups.
-        // We will soon sort, so the initial order does not matter.
         #[allow(rustc::potential_query_instability)]
-        let mut groups: Vec<_> = self
+        let groups: Vec<_> = self
             .lint_groups
             .iter()
             .filter_map(|(k, LintGroup { depr, .. })| depr.is_none().then_some(k))
             .collect();
-        groups.sort();
         let groups = groups.iter().map(|k| Symbol::intern(k));
```

Comment:
This changes behavior! :warning: 

Before, only the groups were sorted and the lints were just appended to the sorted groups. So it would first look at the groups and then at the lints. Now, it will sort groups+lints together, so the groups will be mixed into the lints. 

Please revert those changes to keep the old behavior. With my suggestion below, those changes should no longer be necessary.